IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m

```diff
@@ -26,21 +26,375 @@
+        {
+            response = [MSIDWebviewAuthorization responseWithURL:url
+                                                    requestState:self.requestState
+                                                   stateVerifier:self.stateVerifier
+                                                         context:self.context
+                                                           error:&systemError];
+        }
+        
+        dispatch_async( dispatch_get_main_queue(), ^{
+            completionHandler(response, error);
+        });
+    }
+    
+    [_completionLock unlock];
+}
+
+- (void)startRequest:(NSURLRequest *)request
+{
+    [self loadRequest:request];
+    [self presentView];
+}
+
+- (void)loadRequest:(NSURLRequest *)request
+{
+    [self.webView loadRequest:request];
+}
+
+#pragma mark - WKNavigationDelegate Protocol
+
+- (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigationAction *)navigationAction decisionHandler:(void (^)(WKNavigationActionPolicy))decisionHandler
+{
+    NSURL *requestUrl = navigationAction.request.URL;
+    NSString *requestUrlString = [requestUrl.absoluteString lowercaseString];
+    
+    MSID_LOG_VERBOSE(self.context, @"-decidePolicyForNavigationAction host: %@", [MSIDAuthority isKnownHost:requestUrl] ? requestUrl.host : @"unknown host");
+    MSID_LOG_VERBOSE_PII(self.context, @"-decidePolicyForNavigationAction host: %@", requestUrl.host);
+    
+    // Stop at the end URL.
+    if ([requestUrlString hasPrefix:[_endUrl.absoluteString lowercaseString]] ||
+        [[[requestUrl scheme] lowercaseString] isEqualToString:@"msauth"])
+    {
+        self.complete = YES;
+        
+        NSURL *url = navigationAction.request.URL;
+        [self webAuthDidCompleteWithURL:url];
+        
+        decisionHandler(WKNavigationActionPolicyCancel);
+        return;
+    }
+    
+    if ([requestUrlString isEqualToString:@"about:blank"])
+    {
+        decisionHandler(WKNavigationActionPolicyAllow);
+        return;
+    }
+    
+    if ([[[requestUrl scheme] lowercaseString] isEqualToString:@"browser"])
+    {
+        self.complete = YES;
+        requestUrlString = [requestUrlString stringByReplacingOccurrencesOfString:@"browser://" withString:@"https://"];
+        
+#if TARGET_OS_IPHONE
+        if (![MSIDAppExtensionUtil isExecutingInAppExtension])
+        {
+            [self cancel];
+            [MSIDAppExtensionUtil sharedApplicationOpenURL:[[NSURL alloc] initWithString:requestUrlString]];
+        }
+        else
+        {
+            MSID_LOG_INFO(self.context, @"unable to redirect to browser from extension");
+        }
+#else
+        [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:requestUrlString]];
+#endif
+        
+        decisionHandler(WKNavigationActionPolicyCancel);
+        return;
+    }
+    
+#if TARGET_OS_IPHONE
+    // check for pkeyauth challenge.
+    if ([requestUrlString hasPrefix:[kMSIDPKeyAuthUrn lowercaseString]])
```

Comment:
again, AAD specific logic shouldn't be handled here

Reply:
addressed by having a subclass MSIDAADOAuthEmbeddedWebviewController.